src/Compilers/Core/Portable/NativePdbWriter/PdbWriter.cs

```diff
@@ -31,6 +35,161 @@ internal sealed class PdbWritingException : Exception
+            _hashAlgorithm = logging ? new SHA1CryptoServiceProvider() : null;
+        }
+
+        private void Flush()
+        {
+            if (_logData.BaseStream.Length > 0)
+            {
+                _hashAlgorithm.TransformBlock(_logData.BaseStream.Buffer, (int)_logData.BaseStream.Position);
+                _logData.BaseStream.Position = 0;
+            }
+        }
+
+        private void MaybeFlush()
+        {
+            if (_logData.BaseStream.Length >= bufferFlushLimit)
+            {
+                _hashAlgorithm.TransformBlock(_logData.BaseStream.Buffer, (int)_logData.BaseStream.Position);
+                _logData.BaseStream.Position = 0;
+            }
+        }
+
+        internal ContentId ContentIdFromLog()
+        {
+            Debug.Assert(_logData.BaseStream != null);
+            _hashAlgorithm.TransformFinalBlock(_logData.BaseStream.Buffer, (int)_logData.BaseStream.Position);
+            _logData.BaseStream.Position = 0;
+            return ContentId.FromHash(_hashAlgorithm.Hash.ToImmutableArray());
+        }
+
+        internal void Close()
+        {
+            _hashAlgorithm?.Dispose();
+            _logData.BaseStream?.Free();
+        }
+
+        internal enum PdbWriterOperation : byte
+        {
+            SetUserEntryPoint,
+            DefineDocument,
+            SetCheckSum,
+            OpenMethod,
+            OpenScope,
+            CloseMethod,
+            CloseScope,
+            UsingNamespace,
+            DefineSequencePoints,
+            SetSymAttribute,
+            DefineConstant2,
+            DefineLocalVariable2,
+            DefineAsyncStepInfo,
+            DefineCatchHandlerILOffset,
+            DefineKickoffMethod,
+            OpenMapTokensToSourceSpans,
+            MapTokenToSourceSpan,
+            CloseMapTokensToSourceSpans
+        }
+
+        public bool LogOperation(PdbWriterOperation op)
+        {
+            var logging = this._logging;
+            if (logging)
+            {
+                LogArgument((byte)op);
+            }
+
+            return logging;
+        }
+
+        public void LogArgument(uint[] data)
+        {
+            _logData.WriteInt(data.Length);
+            for (int i = 0; i < data.Length; i++)
+            {
+                _logData.WriteUint(data[i]);
+            }
+            MaybeFlush();
+        }
+
+        public void LogArgument(string data)
+        {
+            _logData.WriteString(data);
+            MaybeFlush();
+        }
+
+        public void LogArgument(uint data)
+        {
+            _logData.WriteUint(data);
+        }
+
+        public void LogArgument(byte data)
+        {
+            _logData.WriteByte(data);
+        }
+
+        public void LogArgument(byte[] data)
+        {
+            LogArgument(data.Length);
+            _logData.WriteBytes(data);
+            MaybeFlush();
+        }
+
+        public void LogArgument(int[] data)
+        {
+            LogArgument(data.Length);
```

Comment:
Is there a real need to log length? Should it be in bytes or should the type of the elements be reflected in the log?